### PR TITLE
proposed msgpack 0.6.0 fix

### PR DESCRIPTION
**Pull request by tjb900**
Possible fix for #2392

### CODE REVIEW

**Comment by mrocklin** [comment]
Thanks for the fix @tjb900 !

A couple small comments:

-  It looks like the flake8 linter didn't like one of the lines here:

        distributed/protocol/serialize.py:61:30: E226 missing whitespace around arithmetic operator

-  Can you think of a good test that we can add to ensure that regressions don't occur in the future?

**Comment by tjb900** [comment]
@mrocklin Thanks for the feedback.  Have fixed up a couple of things, and added a test which fails before the fix (although it seems unlikely that a future regression would hit exactly that particular scenario).

AFAICT the remaining test failures seem to be transient issues in the CI?

File: distributed/protocol/tests/test_protocol.py [hunk 1]
```diff
@@ -137,6 +137,23 @@ def test_large_messages():
     assert (msg['y']['a'].data == msg2['y']['a']).all()
 
 
+@slow
+def test_large_messages_map():
+    np = pytest.importorskip('numpy')
```
Comment (mrocklin) [inline distributed/protocol/tests/test_protocol.py]: Do we need the numpy check here?

File: distributed/protocol/tests/test_protocol.py [hunk 2]
```diff
@@ -137,6 +137,23 @@ def test_large_messages():
     assert (msg['y']['a'].data == msg2['y']['a']).all()
 
 
+@slow
```
Comment (mrocklin) [inline distributed/protocol/tests/test_protocol.py]: This test seems to run quickly enough for me.

```python
In [1]: from distributed.protocol.tests.test_protocol import dumps, loads                                          

In [2]: %time x = {i: 'mystring_%d' % i for i in range(100000)}                                                    
CPU times: user 37.7 ms, sys: 3.42 ms, total: 41.2 ms
Wall time: 40.6 ms

In [3]: %time b = dumps(x)                                                                                         
CPU times: user 40.9 ms, sys: 99 µs, total: 41 ms
Wall time: 40.7 ms

In [4]: x2 = loads(b)                                                                                              

In [5]: %time x2 = loads(b)                                                                                        
CPU times: user 35.4 ms, sys: 324 µs, total: 35.7 ms
Wall time: 34.7 ms

In [6]: %time assert x == x2                                                                                       
CPU times: user 4.94 ms, sys: 97 µs, total: 5.04 ms
Wall time: 5.08 ms
```

We can probably drop the `@slow` decorator

File: distributed/protocol/tests/test_protocol.py [hunk 3]
```diff
@@ -137,6 +137,23 @@ def test_large_messages():
...
+def test_large_messages_map():
+    np = pytest.importorskip('numpy')
+    psutil = pytest.importorskip('psutil')
```
Comment (mrocklin) [inline distributed/protocol/tests/test_protocol.py]: It looks like we now depend on `psutil`, so we can probably drop this.

File: distributed/protocol/tests/test_protocol.py [hunk 4]
```diff
@@ -137,6 +137,23 @@ def test_large_messages():
...
+    if psutil.virtual_memory().total < 8e9:
+        return
+
+    if sys.version_info.major == 2:
+        return 2
```
Comment (mrocklin) [inline distributed/protocol/tests/test_protocol.py]: I'm curious, what doesn't work about this test in Python 2?

**Comment by mrocklin** [comment]
And I agree that those failures look unrelated.  We've been plagued by
these recently.  I'll take a look at them later today.

**Comment by tjb900** [comment]
Fixed.  Most of these were due to lazy copying from the test above.
Have also changed it to explicitly skip (rather than silently pass) in the low-memory situation.

**Comment by mrocklin** [comment]
This looks good to me.  Thank you for resolving this @tjb900 

Also, I notice that this is your first commit to this repository.  Welcome!

**Comment by rsignell-usgs** [comment]
And chance to get a new minor release to include this fix?  This broke [one of the pangeo workshop demos ](https://github.com/pangeo-data/pangeo-tutorial-agu-2018/issues/12) here at AGU and it would be nice to fix it with a `dask` update rather than a `msgpack` rollback.

**Pull request merged**